src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs

```diff
@@ -46,96 +42,174 @@ public class Settings : ISettings
+            {
+                return false;
+            }
+
+            var outputSettingsFile = GetOutputSettingsFileForSection(section.Name);
+
+            if (outputSettingsFile != null)
+            {
+                return outputSettingsFile.RootElement.AddChild(section, isBatchOperation);
+            }
+
+            return false;
         }
 
-        public Settings(string root, string fileName, bool isMachineWideSettings)
+        public bool SetItemInSection(string sectionName, SettingsItem item, bool isBatchOperation = false)
         {
-            if (string.IsNullOrEmpty(root))
+            if (string.IsNullOrEmpty(sectionName))
             {
-                throw new ArgumentException("root cannot be null or empty", nameof(root));
+                throw new ArgumentNullException(nameof(sectionName));
             }
 
-            if (string.IsNullOrEmpty(fileName))
+            if (item == null)
             {
-                throw new ArgumentException(Resources.Argument_Cannot_Be_Null_Or_Empty, nameof(fileName));
+                throw new ArgumentNullException(nameof(item));
             }
 
-            if (!FileSystemUtility.IsPathAFile(fileName))
+            // Check if set is an update
+            if (_computedSections.TryGetValue(sectionName, out var section))
             {
-                throw new ArgumentException(Resources.Settings_FileName_Cannot_Be_A_Path, nameof(fileName));
+                if (section.TryUpdateChildItem(item, isBatchOperation))
+                {
+                    return true;
+                }
             }
 
-            Root = root;
-            FileName = fileName;
-            XDocument config = null;
-            ExecuteSynchronized(() => config = XmlUtility.GetOrCreateDocument(CreateDefaultConfig(), ConfigFilePath));
-            ConfigXDocument = config;
-            IsMachineWideSettings = isMachineWideSettings;
-            CheckConfigRoot();
+            var outputSettingsFile = GetOutputSettingsFileForSection(sectionName);
+
+            // Set is an add
+            return outputSettingsFile != null && outputSettingsFile.RootElement.AddItemInSection(sectionName, item.Copy(), isBatchOperation);
         }
 
         public event EventHandler SettingsChanged = delegate { };
 
-        public IEnumerable<ISettings> Priority
+        public Settings(string root)
+            : this(new SettingsFile(root)) { }
+
+        public Settings(string root, string fileName)
+            : this(new SettingsFile(root, fileName)) { }
+
+        public Settings(string root, string fileName, bool isMachineWide)
+            : this(new SettingsFile(root, fileName, isMachineWide)) { }
+
+        private Settings(SettingsFile settingsHead)
+        {
+            _settingsHead = settingsHead;
+
+            var curr = _settingsHead;
+            while (curr != null)
+            {
+                curr.SettingsChanged += (_, __) => { RefreshSettingValues(); };
```

Comment:
do you really need to rely on `SettingsChanged` event from `SettingFile` to take latest changes in `ISettings`? If I'm not wrong then every change in any of the `SettingFile` happens through `ISettings`...

Reply:
As I just focused this refactor in getting the same behavior as we had previously, then all the changes happen from `ISettings`, but this is not true for the whole API. The API has the capability for users only editing one `ISettingsFile` directly (e.g. if we add the ability to edit specific configs in the UI), therefore I made it so that if there is any change with any of the `ISettingsFile` used in a given `ISettings` then we recompute the merged values.

The previous code calculated the merged values in every `Get` operation, I updated this to make `get` more performant (since we do way more gets than sets), and only recompute the values on each `set`. Therefore I would say that this code is way more performant than the previous.

Reply:
Even if we allow editing individual config file through UI, that will also be through `Settings` or would we allow editing multiple files at the same time? 

I think we can take all this discussion offline and finalize it...?